processing/src/main/java/io/druid/segment/QueryableIndexStorageAdapter.java

```diff
@@ -310,7 +331,10 @@ public void advance()
                       if (Thread.interrupted()) {
                         throw new QueryInterruptedException();
                       }
-                      cursorOffset.increment();
+
+                      do {
+                        cursorOffset.increment();
+                      } while (!isDone() && !filterMatcher.matches());
```

Comment:
If we change `Cursor#advance()` and `Offset#increment()` return boolean, we can remove `isDone()` or `withinBounds()`. But I'm not sure it will make a difference in performance.